Add MAX and MIN memory option

HerrMagiic opened the pull request:
**Pull Request Title:** Add environment variables for configuring Minecraft server memory usage

**Pull Request Description:**
This pull request adds two new environment variables to the Dockerfile, providing the ability to configure the maximum and minimum memory usage for a Minecraft server. This feature allows users to easily adjust the memory allocation based on their server's requirements.

**Changes Made:**
- Added two new environment variables, `MEMORYSIZEMIN` and `MEMORYSIZEMAX`, to the Dockerfile.
- Updated the Dockerfile to utilize the new environment variables in the appropriate sections.
- Documented the usage of the new environment variables in the Dockerfile comments and in the README.md.

**Usage Example:**
To set the minimum memory usage to 1GB and the maximum memory usage to 4GB, you can modify the environment variables in your docker-compose.yml file as follows:

```yaml
version: "3.9"
services:
  minecraft:
    image: "marctv/minecraft-papermc-server:latest"
    restart: always
    container_name: "mcserver"
    environment:
      - MEMORYSIZEMIN=1G
      - MEMORYSIZEMAX=4G
```

Please review and merge this pull request. Thank you!

NilsSchucka committed:
Add MAX and MIN memory option

mtoensing commented:
Hi, your support is much appreciated. May I ask why this is necessary? You can limit the memory already and PaperMC or Minecraft in general loves a fixed memory pool. So why a second and third argument to have a max and min value? And if we need a min, why can the max be the existing parameter and we only add a second min parameter? Sorry for the questions. I really would like to understand the problem for this solution.

HerrMagiic commented:
Hey, sorry for the late reply.

I submitted this pull request because I thought it would be a good idea to add the feature of setting a minimum memory allocation. The reason behind this addition is that I don't want my Minecraft server to constantly use the assigned memory. If there is any available memory, it should be freed for my other services on my server, and maybe other user of this container would like this option too🤷‍♂️. 

And yes, I think you're right, it would be better to just add the minmemory parameter and leave the max param as it is. I just changed the param to max for a better understanding.

I added this for my personal use and I need to say I have little to no idea about what's better for a minecraft server in terms of memory and performance.

mtoensing commented:
_I added this for my personal use and I need to say I have little to no idea about what's better for a minecraft server in terms of memory and performance._

From my experience, we should stick to max https://docs.papermc.io/velocity/tuning. Because what happens if there is not enough memory available if the server uses the min value and another process on the server uses the memory. I understand that it feels wrong that the server allocates xxx GB even if no player is online. But that changes within seconds of a player comes online. So, I would propose that we leave it as it is. But this is open for discussion.

razaqq commented:
I would disagree here, more configuration is better. You dont know the usecase of each user. What if someone runs a server for another game on the same machine thats popular in a different timezone and needs more ram then? Just because you dont need it doesnt mean someone else wont.

Ideally you just add a JAVA_ARGS env var and make MEMORYSIZE optional, so that users have complete freedom over what they want.

mtoensing commented:
@razaqq but you can configure the memorysize. There is just no "min" because from my experience, a minimum memory does not make sense. Just as I explained in my last comment with a link to the documentation of PaperMC.

razaqq commented:
Again, just because you dont need it, doesnt mean nobody else needs it. There are other java arguments someone might want to provide, not just memory related.

Also your link goes to some server proxy, i have no idea what that is supposed to have to do with anything.

mtoensing commented:
Sorry, I meant this link: https://docs.papermc.io/paper/aikars-flags 

But i don't get the problem. You can always override the java_flags and papermc_flags and add everything you want manually with this container. But the main flag "memorysize" should stay simple because I need a good explanation why we should ditch the easy to understand memorysize variable for a min and max value if: 

a) you can do this manually
b) a server should not need a minimum value because it performs better if you have a fixed size of allocated ram.

I added the JAVA_FLAGS parameter to the documentation. It was missing.

razaqq commented:
Right thats exactly what i meant.

However when JAVAFLAGS is set that overrides the default java args right? So when it is set, -Xms and -Xmx shouldnt be set either.

mtoensing commented:
"_However when JAVAFLAGS is set that overrides the default java args right? So when it is set, -Xms and -Xmx shouldnt be set either._"

I think that when JAVAFLAG sets -Xms and Xmx that should override the previous set values.

"In Java command-line parameters, the order of appearance does matter. When you have conflicting options, the last occurrence will generally take precedence. So, if you set -Xms and -Xmx values in both $MEMORYSIZE and $JAVAFLAGS, the values from $JAVAFLAGS will override the values from $MEMORYSIZE.

Here's a breakdown of what happens:

The command-line starts processing from left to right.
-Xms$MEMORYSIZE is encountered first, and it sets the initial heap size.
-Xmx$MEMORYSIZE is encountered next, and it sets the maximum heap size. Since -Xmx appears after -Xms, it will override the previously set initial heap size from -Xms.
If you have both -Xms and -Xmx values specified in both places ($MEMORYSIZE and $JAVAFLAGS), the values in $JAVAFLAGS will be the ones that take effect because they are encountered later in the command-line."

razaqq commented:
Yeah i read the dockerfile. Should add a small mention of JAVAFLAGS overriding everything in the docs.

mtoensing commented:
Done.
